### PR TITLE
Spark: Add CopyTable spark action

### PR DESCRIPTION
**Pull request by laithalzyoud**
This PR adds a new Spark action to copy an Iceberg table. The action includes processing metadata files, manifest lists, and position delete files to reflect changes in the location prefixes, aiming to support operations like migration to a new storage location or table duplication.

Here's a breakdown of what it does and how it works:

- Initializes and validates input parameters, such as source and target location prefixes, start and end versions.
- Executes the copy action, which involves rebuilding metadata to reflect the new locations, creating a staging area for the copied files, and generating lists of data and metadata files to move.
- Supports rewriting of metadata files, manifest files and lists, and position delete files to update paths inside these files to accommodate the new location prefix.
- Identify which data files need to be moved based on the snapshots included between the specified start and end versions of the table.
- Utilizes Spark to parallelize the processing of rewriting files, enabling efficient handling of large tables.
- Supports Spark 3.3, 3.4 and 3.5

This PR extends @flyrain original PR https://github.com/apache/iceberg/pull/4705

### CODE REVIEW

**Comment by manuzhang** [comment]
How about position delete files?

**Comment by laithalzyoud** [comment]
> How about position delete files?

@manuzhang They are covered in this PR, let me add that in the description as well 👌

**Comment by nastra** [comment]
> Supports Spark 3.3, 3.4 and 3.5

We should probably first focus on a single Spark version (3.5) and once the PR is merged, backport the changes to previous Spark versions. Otherwise it will be difficult to review/change the same stuff across multiple Spark versions.

File: core/src/main/java/org/apache/iceberg/actions/BaseCopyTableActionResult.java
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/actions/BaseCopyTableActionResult.java:21]: this should probably be similar to how all the other Result classes are implemented, such as https://github.com/apache/iceberg/blob/82e0a56323537432e087a07b4cfd5100bf37073d/core/src/main/java/org/apache/iceberg/actions/BaseDeleteReachableFiles.java#L23-L33

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestCopyTableAction.java
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestCopyTableAction.java:511]: this is deprecated code. please use `assertThatThrownBy(...).isInstanceOf(..).hasMessage(...)`

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestCopyTableAction.java
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestCopyTableAction.java:767]: actual/expected are wrong here. Should be `assertThat(filesToMove).hasSize(count)`

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestCopyTableAction.java
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestCopyTableAction.java:753]: actual/expected are wrong here and should be the other way around

File: spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestCopyTableAction.java
Comment (nastra) [inline spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestCopyTableAction.java:780]: it seems a bunch of assertions have actual/expected in the wrong order. Please also update all the other places

File: core/src/main/java/org/apache/iceberg/TableMetadataUtil.java
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/TableMetadataUtil.java:33]: please add a `TestTableMetadataUtil` with some tests where metadata/prefixes can be null/empty/invalid/valid

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/BaseCopyTableSparkAction.java
Comment (nastra) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/BaseCopyTableSparkAction.java:136]: it's not clear what `sVersion` refers to, so why not `newStartVersion`? Same for all the other params

**Comment by nastra** [comment]
@laithalzyoud thanks for working on this. I just did a very quick high-level review, but will do a more thorough one this week

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/BaseCopyTableSparkAction.java
Comment (jotarada) [inline spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/BaseCopyTableSparkAction.java:380]: Could be rewritten as ```List<MetadataLogEntry> versions = Lists.reverse(metadata.previousFiles());
 for (MetadataLogEntry version: versions) {
      if (version.file().equals(startVersion)) {
        break;
      }
    }```

**Review by amogh-jahagirdar — CHANGES_REQUESTED** [review]
Thanks @laithalzyoud this is great to see! Beyond @nastra's point of doing Spark 3.5 separately, it would be ideal to have the `CopyTable` API changes be in a separate PR first.

Having the API changes be separate allows us to discuss things like semantics/expectations/preconditions of the API just so the community is on the same page as to what they can expect when working with this action and if the right options are exposed.

After that point, we can look at the implementation. One aspect of implementation that I think I'd also separate is the exposing of `ManifestLists`/`ManifestReader`. I totally get why that's required for this action but I think it's worth having that in a separate commit.

Lastly, this gets more into the API and implementation but I think we should figure out if there should be separate sorts of exposed "operations" for replicating a given manifest/manifest list rather than having it all embedded in the spark procedure etc. Those operations can invidvidually be used for reasons beyond copying for replication; they can be used for fixing corrupt metadata if the right APIs are exposed.

**Comment by ajantha-bhat** [comment]
@laithalzyoud : Are you planning to address the comments on this? This feature is definitely useful. 
If not, I would like to take it up.

**Comment by laithalzyoud** [comment]
> @laithalzyoud : Are you planning to address the comments on this? This feature is definitely useful. If not, I would like to take it up.

Hey @ajantha-bhat! Yes I'm planning to continue working on this soon, you can help in the code review if you'd like 👍

**Comment by huaxingao** [comment]
@laithalzyoud Thanks for your work on this PR! I've noticed there hasn't been activity for a while, and I wanted to check if you're still able to continue working on it. If you're busy with other commitments and would like some help, I’d be glad to take over or assist. Thanks!

**Comment by moomindani** [comment]
@laithalzyoud I am super interested in this PR and it will unblock many use cases. Are you working on this now?

**Comment by huaxingao** [comment]
@laithalzyoud Thanks for the thumbs-up! Could you please confirm if you are planning to continue working on this PR, or would you like me to take over? I’m happy to help in any way needed. Thank you!

File: spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/BaseCopyTableSparkAction.java
Comment (laithalzyoud) [reply spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/BaseCopyTableSparkAction.java:358]: If we want to support actually moving the files - then we will need to support different cloud providers (GCP, AWS, Azure) as well as on-prem setups (i.e copying in Linux). It might not be ideal for some use-cases as well - for example in our case where we are actually using this in a production setting, copying using GCS client libraries was extremely slow and inefficient for huge tables (>5TB) and instead we used a [managed service from GCP](https://cloud.google.com/storage-transfer/docs/cloud-storage-to-cloud-storage ) to handle the move efficiently. So this can actually be very specific to where the files are stored and where you want to move them, you can consider a use-case of someone migrating between 2 different cloud providers or moving data from on-premise to the cloud for example and so on. So from my perspective for it's better to just rewrite the paths so the table is usable in a new location and leave the actual copying to the users, maybe in a future iteration some basic interface to move files for common use-cases can be implemented

**Comment by laithalzyoud** [comment]
Hey @huaxingao! I'm planning to continue working on it starting this week. For now I'll close this PR and open a new one to just add the interface, once we agree on the interface, I'll create the implementation PR after like agreed earlier with @flyrain and @amogh-jahagirdar  👍

**Comment by laithalzyoud** [comment]
I created the [PR](https://github.com/apache/iceberg/pull/10920) to just add the interface, please feel to review it and provide feedback!

**Comment by loudwanderingdune** [comment]
I see the interface PR #10920 has been merged. Is the implementation ready to be worked on?

**Comment by manuzhang** [comment]
@loudwanderingdune This feature is already implemented and released. Please check out https://iceberg.apache.org/docs/nightly/spark-procedures/#table-replication.

**Pull request closed without merging**
